In [6]:
from pathlib import Path
parent_dir = str(Path.cwd().parent)
%cd $parent_dir

D:\OrenRichter\Research\pyERGM


D:\OrenRichter\Research\pyERGM\venv\Lib\site-packages\IPython\core\magics\osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


In [8]:
from utils import *
from ergm import ERGM

import pandas as pd
import numpy as np

### Initialize an ERGM with random coefficients

In [9]:
n_nodes = 10
stats_calculator = NetworkStatistics(metric_names=["num_edges"], )


stats = [NumEdges(), CustomMetric(lambda x: ), ]
ergm = ERGM(n_nodes, stats_calculator, is_directed=False)
ergm = ERGM(n_nodes, stats_calculator, is_directed=True)

ergm.print_model_parameters()

Number of nodes: 10
Thetas: [0.33527386]
Normalization factor approx: 36.49400038899532
Is directed: False


### Sample networks from the distribution, and calculate some statistics on the distribution

In [10]:
network_stats = []
for i in range(10):
    W = ergm.sample_network(sampling_method="NaiveMetropolisHastings", steps=1000)
    G = connectivity_matrix_to_G(W, directed=False)
    edge_count = len(G.edges())
    triangle_count = sum(nx.triangles(G).values()) // 3

    network_stats.append({"edge_count": edge_count, "triangle_count": triangle_count})

network_stats_df = pd.DataFrame(network_stats)
network_stats_df.head(20)


edge_count  triangle_count
0          45             120
1          45             120
2          45             120
3          45             120
4          45             120
5          45             120
6          45             120
7          45             120
8          45             120
9          45             120

## Fit an ERGM

We begin by initializing an ERGM with predefined parameters - 

In [11]:
n_nodes = 5
stats_calculator = NetworkStatistics(metric_names=["num_edges"])

theta = -np.log(3)
ergm = ERGM(n_nodes, stats_calculator, is_directed=False, initial_thetas=[theta])

print("Baseline ERGM parameters - ")
ergm.print_model_parameters()


Baseline ERGM parameters - 
Number of nodes: 5
Thetas: [np.float64(-1.0986122886681098)]
Normalization factor approx: 43.441946503045074
Is directed: False


We now sample from this model a network that will become our "observed network", when estimating our fit performance.

In [12]:
W = ergm.sample_network(sampling_method="NaiveMetropolisHastings", steps=1000)

G = connectivity_matrix_to_G(W, directed=False)

real_edge_count = len(G.edges())
real_triangle_count = sum(nx.triangles(G).values()) // 3

print(f"Sampled a random network from a model with theta = {theta}")
print(W)
print(f"Network has statistics - edge count: {real_edge_count}, triangle count: {real_triangle_count}")

Sampled a random network from a model with theta = -1.0986122886681098
[[0. 0. 1. 1. 1.]
 [0. 0. 0. 1. 0.]
 [1. 0. 0. 0. 0.]
 [1. 1. 0. 0. 0.]
 [1. 0. 0. 0. 0.]]
Network has statistics - edge count: 4, triangle count: 0


Now, fit a random ERGM to create networks with similar statistics

In [13]:
fitted_ergm = ERGM(n_nodes, stats_calculator, is_directed=False)
print(f"Initial ERGM parameters:")
fitted_ergm.print_model_parameters()

pre_fit_W_matrices = []
samples_before_fit = []

n_samples = 10

for i in range(n_samples):
    sampled_W = fitted_ergm.sample_network(sampling_method="NaiveMetropolisHastings", steps=1000)
    G = connectivity_matrix_to_G(sampled_W, directed=False)
    edge_count = len(G.edges())
    triangle_count = sum(nx.triangles(G).values()) // 3

    samples_before_fit.append({"edge_count": edge_count, "triangle_count": triangle_count})
    pre_fit_W_matrices.append(sampled_W)

samples_before_fit_df = pd.DataFrame(samples_before_fit)
print("")
# Print mean for each metric - 
print(f"Mean of samples before fitting, calculated on {n_samples} samples- ")
print(samples_before_fit_df.mean())


Initial ERGM parameters:
Number of nodes: 5
Thetas: [0.38415461]
Normalization factor approx: 50.33647025496875
Is directed: False

Mean of samples before fitting, calculated on 10 samples- 
edge_count        10.0
triangle_count    10.0
dtype: float64


We now fit the ERGM!

In [14]:

print(f"Fitting ERGM...")
fitted_ergm.fit(W, optimization_options={'disp': True, 'maxiter': 500, 'maxfev': 500}, n_networks_for_norm=500, n_mcmc_steps=100)
print("Done fitting!")


Fitting ERGM...
	Starting fit with initial normalization factor: 50.33647025496875
	Optimization options: {'disp': True, 'maxiter': 500, 'maxfev': 500}
	Optimization result:
	Theta: [-0.39642108]
	Normalization factor: 41.88198286288824
       message: Maximum number of function evaluations has been exceeded.
       success: False
        status: 1
           fun: 5.243735736850018
             x: [-3.964e-01]
           nit: 189
          nfev: 500
 final_simplex: (array([[-3.964e-01],
                       [-3.964e-01]]), array([ 5.244e+00,  5.271e+00]))
Done fitting!


D:\OrenRichter\Research\pyERGM\ergm.py:129: RuntimeWarning: Maximum number of function evaluations has been exceeded.
  # OR: Providing the Scipy optimizer with a gradient function can streamline the optimization substantially. The


In [48]:
print(f"Post fit ERGM parameters:")
fitted_ergm.print_model_parameters()

post_fit_W_matrices = []
samples_after_fit = []

for i in range(n_samples):
    sampled_W = fitted_ergm.sample_network(sampling_method="NaiveMetropolisHastings", steps=1000)
    G = connectivity_matrix_to_G(sampled_W, directed=False)
    edge_count = len(G.edges())
    triangle_count = sum(nx.triangles(G).values()) // 3

    samples_after_fit.append({"edge_count": edge_count, "triangle_count": triangle_count})
    post_fit_W_matrices.append(sampled_W)

samples_after_fit_df = pd.DataFrame(samples_after_fit)
print("")

print(f"Mean of samples after fitting, calculated on {n_samples} samples- ")
print(samples_after_fit_df.mean())



Post fit ERGM parameters:
Number of nodes: 5
Thetas: [-0.15381744]
Normalization factor approx: 135.44816303424523
Is directed: False

Mean of samples after fitting, calculated on 10 samples- 
edge_count        8.4
triangle_count    6.0
dtype: float64


$$
\hat{\theta} = -0.1538 , \theta = -0.3369
\\
$$

edge count and triangle_count not looking better than the random ergm...